File: app/services/comment_service.py
from datetime import datetime

from sqlalchemy.orm import Session

from app.crud import comment_crud
from app.domains.comment import Comment


def create_comment(user_id: int, post_id: int, content: str, db: Session) -> None:
    """
    댓글을 생성합니다.
    TODO:
    """
    comment = Comment(id=None, user_id=user_id, post_id=post_id, content=content, parent_id=None,
                      created_at=datetime.now())
    comment_crud.create_comment(db=db, comment=comment)


def update_comment(user_id: int, comment_id: int, content: str, db: Session) -> Comment:
    """
    댓글을 수정합니다.
    """
    comment = comment_crud.get_comments_by_id(comment_id=comment_id, db=db)
    if comment.user_id != user_id:
        raise ValueError("수정 권한이 없습니다.")
    saved_comment = Comment(
        id=comment.id,
        user_id=comment.user_id,
        content=content,
        post_id=comment.post_id,
        parent_id=comment.parent_id,
        created_at=datetime.now(),
    )
    comment_crud.update_comment(db=db, comment=saved_comment)

File: core/src/main/java/org/cosmic/cafe/application/comment/CommentService.java
package org.cosmic.cafe.application.comment;

import lombok.RequiredArgsConstructor;
import org.cosmic.cafe.application.comment.dto.CommentResponseDTO;
import org.cosmic.cafe.application.member.MemberService;
import org.cosmic.cafe.domain.Comment.Comment;
import org.cosmic.cafe.domain.Comment.CommentRepository;
import org.cosmic.cafe.domain.Comment.exception.CommentErrorCode;
import org.cosmic.cafe.exception.type.BadRequestException;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class CommentService {

    private final CommentRepository commentRepository;
    private final MemberService memberService;

    public UUID saveComment(UUID userId, String content, UUID postId, UUID parentId) {

        Comment comment = createComment(userId, content, postId, parentId);

        return commentRepository.save(comment).getId();
    }

    public void deleteComment(UUID commentId, UUID memberId) {
        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new BadRequestException("해당 댓글이 존재하지 않습니다.", CommentErrorCode.NO_SUCH_COMMENT));
        comment.validateOwner(memberId);
        commentRepository.deleteById(commentId);
    }

    public CommentResponseDTO modifyComment(String content, UUID commentId, UUID memberId) {
        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new BadRequestException("해당 댓글이 존재하지 않습니다.", CommentErrorCode.NO_SUCH_COMMENT));
        Comment modifiedContent = comment.modifyContent(memberId, content);
        commentRepository.save(modifiedContent);
        return new CommentResponseDTO(modifiedContent, memberService.findById(modifiedContent.getUserId()));
    }

    public List<CommentResponseDTO> getCommentsByPostId(UUID postId) {
        List<Comment> comments = commentRepository.findByPostId(postId);
        return comments.stream().map(comment -> new CommentResponseDTO(comment, memberService.findById(comment.getUserId()))).toList();
    }

    public CommentResponseDTO getCommentById(UUID commentId) {
        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new BadRequestException("해당 댓글이 존재하지 않습니다.", CommentErrorCode.NO_SUCH_COMMENT));
        return new CommentResponseDTO(comment, memberService.findById(comment.getUserId()));
    }

    public List<CommentResponseDTO> getCommentByParentId(UUID parentId) {
        List<Comment> comments = commentRepository.findByParentId(parentId);
        return comments.stream().map(comment -> new CommentResponseDTO(comment, memberService.findById(comment.getUserId()))).toList();
    }

    private static Comment createComment(UUID userId, String content, UUID postId, UUID parentId) {
        Comment comment = Comment.builder()
                .content(content)
                .postId(postId)
                .parentId(parentId)
                .userId(userId)
                .createdAt(LocalDateTime.now())
                .build();
        return comment;
    }
}
